### DISCUSSION TITLE
Docker Image

### DISCUSSION THREAD

**Discussion by curtisma** [simple-repository/simple-repository#1] [Ideas]
I think creating a docker image for easily hosting simple-repository-server and simple-repository-browser would be quite helpful.
Would anyone else be intererested in such an image?

**Comment by pelson** [comment]
Sounds like a good idea, and should be fairly simple to do :+1:.
Would you be interested to contribute this to the project? Perhaps via the [github container registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) and an associated GitHub action?

**Reply by curtisma** [reply]
> Would you be interested to contribute this to the project? 

Well, I'm currently looking for a solution to host internal Python packages for some of my team's private projects.  
Are you currently using it?  
How has it been working for you?

> Perhaps via the [github container registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) and an associated GitHub action?


Yeah, definitely, that's how I've done it in the past.

**Reply by pelson** [reply]
Yes, it is powering our internal package repository, which is a combination of internal packages and those on PyPI (some of our internal packages overwrite those on PyPI.org, so it is important that we handle dependency confusion correctly). In our case, we are in an air-gapped environment, so nobody can access PyPI.org directly (they must go through our service).

Some stats: in the last month, we've internally served around 1 million package downloads, to ~600 unique users. Nothing on the scale of PyPI.org, and I don't think we would be able to get close to that scale (e.g. we use sqlite as our db, and want to be able to log each request rather than using a CDN).

It is working well for us - we have found a few crinkles to iron out (e.g. a few months ago, PyPI had a spell of serving responses super slow >20s, and this led to our service stacking up lots of responses and eventually becoming unresponsive), but ultimately we are building out services that rely on modern package repository PEPs, and therefore are motivated to refine simple-repository for our own purposes.

We are actively seeking an answer on whether we should fully openly develop the project - it is mostly a matter of using our resources wisely, and ensuring that we can maximise our broad-scale impact (i.e. across a wide spectrum of activities), so if you get a chance to try it out, any feedback is really valuable to us for that purpose.

I'd be happy to answer any other questions you have - perhaps open new discussions with specific topics, if you have them.

**Reply by pelson** [reply]
FYI, the browser service running at https://simple-repository.app.cern.ch/ has the following Dockerfile:

```
FROM python:3.11.0-alpine

RUN pip install simple-repository-browser==0.4.2 \
;

COPY ./templates /opt/simple-repository-browser/templates

# TODO: Warm up the cache.

ENV XDG_CACHE_DIR=/cache

# Make the cache directory. Note that in OpenShift, this is mounted anyway.
RUN mkdir -p $XDG_CACHE_DIR \
 && chmod a+wr $XDG_CACHE_DIR \
;
CMD simple-repository-browser \
    --port 5000 \
    --templates-dir /opt/simple-repository-browser/templates \
;
```

With a ``templates/base.html`` with content:

```
{% extends "base/base.html" %}

{% block body %}
    <div class="container-fluid text-center" style="background-color: red; color: white; padding: 0.25em;">
      <b>Warning</b>: This is a tech preview of <a href="https://github.com/simple-repository" class="nav-link" style="display: inline;">github.com/simple-repository</a>. The service is unsupported and may be removed at any time.
    </div>
{{ super() }}
{% endblock body %}
```

I decided not to build/run a ``simple-repository-server`` prototype in case somebody started using it heavily, but the ``Dockerfile`` should be comparably straight-forward.
